### PR TITLE
Stops cross-device data movement in tensor iterator

### PR DESCRIPTION
**Pull request by mruberry**
**BC-breaking note:**

In previous versions of PyTorch zero dimensional CUDA tensors could be moved across devices implicitly. For example, 

```
torch.tensor(5, device='cuda:0') + torch.tensor((1, 1), device='cuda:1')
```

would work, even though the tensors are on different CUDA devices. This is a frequent source of user confusion, however, and PyTorch generally does not move data across devices without it being explicit. This functionality is removed in PyTorch 1.6.

**PR Summary:**

Today in PyTorch we allow implicit data movement of zero dimensional CUDA tensors. For example, we allow:

```
torch.tensor(5, device='cuda:0') + torch.tensor((1, 1), device='cuda:1')
```

and 

```
torch.tensor(2, device='cuda') + torch.tensor((3, 5))
```

In both of these cases TensorIterator would move the zero dim CUDA tensor to the device of the non-scalar tensor (cuda:1 in the first snippet, the CPU in the second snippet).

One of PyTorch's fundamental rules, however, is that it does not perform implicit data movement like this, and this change will causes these cases to throw an error. New tests for this behavior are added to test_torch.py, and tests of the old behavior are removed in test_torch.py and test_autograd.py. A cpp test in tensor_iterator_test.cpp is modified to account for the new behavior. 

This addresses https://github.com/pytorch/pytorch/issues/36722.

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Looks good. What is the error message that you get when doing these implicit cross device copies?
I noticed that you don't check for the error message in the tests, just a RuntimeError.

File: test/test_torch.py [hunk 1]
```diff
@@ -13692,30 +13693,56 @@ def run_test(device, dtype):
...
+                with self.assertRaises(RuntimeError):
+                    op(a, cpu_tensor)
+                with self.assertRaises(RuntimeError):
+                    op(cpu_tensor, a)
+
+    # Tests that CPU scalars (including zero dim tensors) can be used in
+    # binary operations with CUDA tensors.
+    @onlyCUDA
+    def test_cuda_cpu_scalar_binary_ops(self, device):
+        t_cuda = torch.randn(1, device=device).clamp(min=1e-3)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py:13726]: A fine idea.

**Comment by mruberry** [comment]
> Looks good. What is the error message that you get when doing these implicit cross device copies?
> I noticed that you don't check for the error message in the tests, just a RuntimeError.

You receive this error message:

```
TORCH_CHECK(false, 
"expected device ", op.device,
" but got device ", op.tensor.device());
```

Which is the same error message you'd receive historically if working with a CUDA tensor that wasn't zero-dim.

**Comment by albanD** [comment]
Ok sounds good, can you make sure in the test that this is this error that is caught (and not another unrelated one) using a `assertRaisesRegex()`?

**Comment by mruberry** [comment]
> Ok sounds good, can you make sure in the test that this is this error that is caught (and not another unrelated one) using a `assertRaisesRegex()`?

Sure.

File: test/test_torch.py [hunk 2]
```diff
@@ -13692,30 +13693,56 @@ def run_test(device, dtype):
...
+        val_tensor = torch.tensor(val_scalar)
+        for op in (operator.add, torch.add,
+                   operator.sub, torch.sub,
+                   operator.mul, torch.mul,
+                   operator.truediv, torch.true_divide,
+                   operator.floordiv, torch.floor_divide):
+            for tensor_val in (1, (1,)):
+                t_cuda = torch.tensor(tensor_val, device=device)
+                t_cpu = t_cuda.cpu()
+            for val in (val_scalar, val_tensor):
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py:13738]: You are correct. My mistake.

[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py:13738]: Correct - looks like some final formatting didn't get saved. You're right on both accounts, of course.

**Review by ngimel — APPROVED** [review]
